packages/plugin-core/src/services/NoteSyncService.ts

```diff
@@ -122,86 +100,26 @@ export class NoteSyncService implements INoteSyncService {
       calculateHash: true,
     });
     note = NoteUtils.hydrate({ noteRaw: note, noteHydrated: oldNote });
-    note = await this.syncNoteMetadata({ note, fmChangeOnly });
-
-    const now = NoteUtils.genUpdateTime();
-    note.updated = now;
+    note = await NoteUtils.updateNoteMetadata({
+      note,
+      fmChangeOnly,
+      engine: this._extension.getEngine(),
+      enableLinkCandidates:
+        this._extension.workspaceService?.config.dev?.enableLinkCandidates,
+    });
 
-    const noteClean = await engine.updateNote(note);
     this.L.debug({ ctx, fname: note.fname, msg: "exit" });
-    return noteClean;
-  }
-
-  /**
-   * Update note metadata (eg. links and anchors)
-   */
-  public async syncNoteMetadata({
-    note,
-    fmChangeOnly,
-  }: {
-    note: NoteProps;
-    fmChangeOnly: boolean;
-  }) {
-    const engine = this._extension.getEngine();
-    // Avoid calculating links/anchors if the note is too long
-    if (
-      note.body.length > ConfigUtils.getWorkspace(engine.config).maxNoteLength
-    ) {
-      return note;
-    }
-    // Links have to be updated even with frontmatter only changes
-    // because `tags` in frontmatter adds new links
-    const links = await engine.getLinks({ note, type: "regular" });
-    if (!links.data) {
-      throw new DendronError({
-        message: "Unable to calculate the backlinks in note",
-        payload: {
-          note: NoteUtils.toLogObj(note),
-          error: links.error,
-        },
-      });
-    }
-    note.links = links.data;
-
-    // if only frontmatter changed, don't bother with heavy updates
-    if (!fmChangeOnly) {
-      const anchors = await engine.getAnchors({
-        note,
-      });
-      if (!anchors.data) {
-        throw new DendronError({
-          message: "Unable to calculate backlinks in note",
-          payload: {
-            note: NoteUtils.toLogObj(note),
-            error: anchors.error,
-          },
-        });
-      }
-      note.anchors = anchors.data;
-
-      if (this._extension.workspaceService?.config.dev?.enableLinkCandidates) {
-        const linkCandidates = await engine.getLinks({
-          note,
-          type: "candidate",
-        });
-        if (!linkCandidates.data) {
-          throw new DendronError({
-            message: "Unable to calculate the backlink candidates in note",
-            payload: {
-              note: NoteUtils.toLogObj(note),
-              error: linkCandidates.error,
-            },
-          });
-        }
-        note.links = note.links.concat(linkCandidates.data);
-      }
-    }
 
     return note;
   }
 
-  private async onDidSave(document: TextDocument) {
-    const ctx = "NoteSyncService:onDidChange";
+  /**
+   * Exposed for testing only. Callback function for vscode.workspace.OnDidSaveTextDocument
+   * @param document
+   * @returns
+   */
+  public async onDidSave(document: TextDocument) {
```

Comment:
instead of exposing this as public, can we use [[Stubbing private functions or properties|dendron://dendron.docs/dev.process.qa.test#stubbing-private-functions-or-properties]]
 instead?

Reply:
Is there an example to use the stub to test the function directly? I'm having trouble getting this method to be triggered from a vscode event. (someone left a similar note about this issue in `WorkspaceWatcher.test.ts`)

Reply:
In this case you'd have to do something like this:
```ts
const onDidSave = textDocumentService["onDidSave" as keyof TextDocumentService].bind(textDocumentService);
onDidSave(...);
```

This is kind of ugly/brittle though, maybe we can keep this private and follow the "exposing private properties for testing" SOP? (the `__DO_NOT_USE_IN_PROD` wrapper).